### DISCUSSION TITLE
docker container publicly available for anyone

### DISCUSSION THREAD

**Discussion by 11notes** [alangrainger/immich-public-proxy#5] [Feature Requests]
Here you go: https://hub.docker.com/r/11notes/immich-public-proxy

**Comment by alangrainger** [comment]
We already have a docker image, but thank you for taking the time to set one up also.

https://github.com/alangrainger/immich-public-proxy/pkgs/container/immich-public-proxy

```bash
docker pull ghcr.io/alangrainger/immich-public-proxy:latest
```

edit: I have added to Docker Hub now also: https://hub.docker.com/r/alangrainger/immich-public-proxy

**Reply by alangrainger** [reply]
In what way?

**Reply by alangrainger** [reply]
Good suggestion re root. What's the benefit of Alpine as base? It increases the size and includes unnecessary packages.

**Reply by alangrainger** [reply]
Thanks, that's a much more productive conversation when you use your words, rather than just saying "Yeah but yours sucks."

I've implemented those changes.

**Reply by alangrainger** [reply]
Any improvements you can suggest in the JS?

**Reply by bo0tzz** [reply]
> but it does, why sugarcoat it? 

Because it makes you come across as hostile. I understand that is probably not your intention, but it's how this communication style is often received. Saying something sucks is mostly fine if you have actionable feedback to go with it, but coming in with nothing more than "what you did is not good" only serves to make people feel bad.

**Reply by alangrainger** [reply]
These are already implemented:

- **Fixed:** not running as root
- **Fixed:** default proper healthcheck
- **Fixed:** no garbage (no extraneous files copied into the image)
- **Not implemented:** Alpine as base (I don't see any reason to do this?) 
- **Fixed:** I mean you add your entire repo, readme and all to a container, that makes no sense. 
- **Fixed:** I suggest you change your build output and only copy the release direczory containing only the files needed and not all the other trash.
- **Fixed:** Not running as root should be normal. 
- **Fixed:** You also have only one main thread, meaning if your node process dies so does the container.

So what am I missing? Also none of those are related to what might be missing skill-wise in JS.

**Reply by 11notes** [reply]
We are slowly getting there! You still are using the -slim tag for your base image, even though that’s not recommended by the creators of node (again, size does not matter in containers, unless we talk multi-GB containers). You are still not providing the option to provide inline configuration for lightgallery. You have now replaced node with pm2-runtime, something that would not fail the container if the process within can’t be spawned anymore (code or exception issues).  The healtcheck for your app is by default 30s, I doubt your app has such long standing queries that it will not respond within a second or less? You also use node as healtcheck. You are aware that every healthcheck (every 30s) will spawn a node process just to check the URL of your app for a proper response? Simply use curl for that, not node. Also, maybe your /healthcheck should do more than just use “ok”, like actually checking if it can reach immich and more. Like an actual healtcheck, not just a check if the HTTP server is still alive.


As for the js thing, I mean you could use ```proccess.env?.IMMICH_URL || http:// immich-server:2283``` to get the variables and not rely on an .env file but I guess you really need that file 😉. That's what I mean with JS skill issues and co.

**Reply by alangrainger** [reply]
Thanks for the feedback, that's very helpful.


- You still are using the -slim tag for your base image, even though that’s not recommended by the creators of node
	- Per [the recommendations of the creators of Node](https://hub.docker.com/_/node), I read there was a caveat for Alpine: "The main caveat to note is that it does use [musl libc⁠](https://musl.libc.org) instead of [glibc and friends⁠](https://www.etalabs.net/compare_libcs.html), so software will often run into issues depending on the depth of their libc requirements/assumptions." Slim did not have this caveat and as I am (their words) "working in an environment where only the `node` image will be deployed", it seemed like the right one to use.
- You are still not providing the option to provide inline configuration for lightgallery. 
	- I opted not to do this, as YAML is not the best place to write a JSON object. I copied the pattern of including a config file based on what other popular images were doing - examples Caddy, Gitea, Forgejo, Synapse. This file is also only optional for people who want to make changes to lightGallery configuration, and by being a JSON file it matches the lightGallery configuration.
- The healtcheck for your app is by default 30s
	- 30-60 seconds was recommended by Docker and elsewhere when I searched for best practices
- I doubt your app has such long standing queries that it will not respond within a second or less?
	- The timeout is [already 5 seconds](https://github.com/alangrainger/immich-public-proxy/blob/be66b16a08254889b40275764b3debad4dff020d/Dockerfile#L24). I could make this 1 second, but does it matter?
- You also use node as healtcheck. You are aware that every healthcheck (every 30s) will spawn a node process just to check the URL of your app for a proper response? Simply use curl for that, not node. 
	- The reason for this is that `slim` does not include curl/wget. But I just realised I can install that as part of the Dockerfile. Again - this is the first docker image I've ever built, so this stuff is all new to me. I didn't want to use the default, large build, as people may be wanting to deploy to low-space environments, since this is a very lightweight proxy. (Plus, additional unnecessary packages mean additional possibilities for vulnerabilities.)
	- ✅ This is now done 
- Also, maybe your /healthcheck should do more than just use “ok”, like actually checking if it can reach immich and more. Like an actual healtcheck, not just a check if the HTTP server is still alive.
	- ✅ This is now done, great suggestion. I originally had it checking if the service was web-accessible, but apparently that's not best practice as that check should be handled on whatever router service you're using.

**Reply by 11notes** [reply]
- Change [502 to 503](https://github.com/alangrainger/immich-public-proxy/blob/0afc820f7bb66d0677e345be1e53026aa44fc203/app/src/index.ts#L76) because [503](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/503) fits better
- You have a node app, that works on musl and on glibc, musl is superior to glibc, also I recommend using mimalloc as your memory manager for node
- Using [inline config]( https://github.com/11notes/docker-immich-public-proxy/blob/8d54e99de1117ac6f8c6db7231eb34a7e9f51ce1/compose.yaml#L9) in yaml is no problem and is the preferred way instead of mounting files. You have to imagine that containers are floating. Your app is just a proxy, why does it need files to store settings? Best would still be `proccess.env?.IMMICH_URL || http://immich-server:2283`

We are slowly getting there 😉 maybe even so far that you unblock me on Reddit.

**Reply by alangrainger** [reply]
Your help was to tell me "[Your image sucks](https://i.imgur.com/s1aywSS.png)" and then [claim the project as yours](https://old.reddit.com/r/selfhosted/comments/1gk4bue/nginx_config_for_immich_to_safely_share_photo/lvjmdke/?context=10000)?

Don't you have any of your own projects without resorting to claiming others as your own?

**Reply by alangrainger** [reply]
I genuinely appreciate the tips and I wish that you had wanted to work with me to make the original project better. Your approach was to insult me with no further information. It relied on me to keep questioning to find any actionable information.

To me it did not feel like it was in the spirit of open-source collaboration.

**Reply by alangrainger** [reply]
> to my surprise you actually paid attention and listened

It's only a good thing to learn and improve. Any other suggestions I am all ears.

> Now you blocked me on Reddit again. Amazing. You really seem to be scared.

You're [not blocked at my end](https://imgur.com/a/DAflT6r)?

**Reply by bo0tzz** [reply]
@11notes this project is GPL licensed. You cannot just change that to MIT on your fork.

**Reply by 11notes** [reply]
@bo0tzz 
Ah perfect, since it’s not a direct github fork but a copy; I automatically added the MIT license since I’m a strong believer of MIT vs GPL, but of course the license is now GPL, even though MIT is the only true FOSS license 😉.

I mean that’s why immich changed from MIT to AGPLv3 right? Riiiiiiiiiiiiiiiiight? I only do MIT, if I publish code, anyone can do whatever they want with it, even take it closed and make money. Sharing knowledge, not gatekeeping it my friend, maybe one day you can learn that.

**Reply by bo0tzz** [reply]
Thank you for fixing it.

>  Sharing knowledge, not gatekeeping it 

MIT allowing you to make something closed-source doesn't seem very "true FOSS" to me ;) This is exactly why Immich changed to AGPL: The riskiest scenario on a license like MIT is that the we as the team making Immich make it closed source. AGPL prevents that, making sure that the Immich we're building always stays free and open.

As a bit of feedback: You're incredibly condescending in your messages, and it's not nice. I recommend toning that down.

**Reply by bo0tzz** [reply]
> No, it doesn’t because it gives you the possibility to take it closed any time you like and want.

If you mean that we're still able to close the Immich source, no we're not, that's the whole point of choosing AGPL as a legally binding license. Immich has code contributions from well over 400 people, the license could only be changed with permission from all of them which is functionally impossible.

Anyways, if you're continuing to be condescending like this I'm not interested in any further discussion.
